src/main/java/org/ehcache/integrations/shiro/EhcacheShiro.java

```diff
@@ -57,16 +57,40 @@ public V get(K k) throws CacheException {
   public V put(K k, V v) throws CacheException {
     trace("Putting object", k);
 
-    V previousValue = get(k);
-    cache.put(k, v);
+    V previousValue = null;
+
+    while (true) {
+      previousValue = get(k);
```

Comment:
Should use `cache.get(k)` since there is some logic (although minimum) in the `get` impl on this class